src/main/java/hudson/remoting/Engine.java

```diff
@@ -213,133 +296,147 @@ public void run() {
+                    // if the public key is not published then JNLP4-connect will refuse to connect
+                    delegate.add(publicKey);
+                }
+                agentTrustManager.setDelegate(delegate);
+
+                events.status("Handshaking");
+                Socket jnlpSocket = connect(endpoint);
+                Channel channel = null;
+
+                try {
+                    // Try available protocols.
+                    boolean triedAtLeastOneProtocol = false;
+                    for (JnlpProtocolHandler<?> protocol : protocols) {
+                        if (!protocol.isEnabled()) {
+                            events.status("Protocol " + protocol.getName() + " is not enabled, skipping");
                             continue;
                         }
-                        port = con.getHeaderField("X-Hudson-JNLP-Port");
-                        if(con.getResponseCode()!=200) {
-                            if(firstError==null)
-                                firstError = new Exception(salURL+" is invalid: "+con.getResponseCode()+" "+con.getResponseMessage());
-                            continue;
+                        if (jnlpSocket == null) {
+                            jnlpSocket = connect(endpoint);
                         }
-                        if(port ==null) {
-                            if(firstError==null)
-                                firstError = new Exception(url+" is not Jenkins");
+                        if (!endpoint.isProtocolSupported(protocol.getName())) {
+                            events.status("Server reports protocol " + protocol.getName() + " not supported, skipping");
                             continue;
                         }
-                        host = con.getHeaderField("X-Jenkins-JNLP-Host"); // controlled by hudson.TcpSlaveAgentListener.hostName
-                        if (host == null) host=url.getHost();
-                        String names = con.getHeaderField("X-Jenkins-Agent-Protocols");
-                        if (names != null) {
-                            agentProtocolNames = new HashSet<String>();
-                            for (String name: names.split(",")) {
-                                name = name.trim();
-                                if (!name.isEmpty()) {
-                                    agentProtocolNames.add(name);
+                        triedAtLeastOneProtocol = true;
+                        events.status("Trying protocol: " + protocol.getName());
+                        try {
+                            channel = protocol.connect(jnlpSocket, headers, new JnlpConnectionStateListener() {
+
+                                @Override
+                                public void beforeProperties(@Nonnull JnlpConnectionState event) {
+                                    if (event instanceof Jnlp4ConnectionState) {
```

Comment:
🐜 So it leaks JNLP4 implementation to base classes. Does not seem to be very good from remoting extensibility PoV. It would be preferable to add handling logic to the protocol class via new method